### PR TITLE
do real close when close server

### PR DESCRIPTION
**Pull request by ngaut**

### CODE REVIEW

File: client.go [hunk 1]
```diff
@@ -28,19 +30,29 @@ type Client struct {
...
+		RemoteAddr:  conn.RemoteAddr().String(),
+		conn:        conn,
 	}
 }
 
-// Close will disconnect the client when the buffer has been send
-func (i *Client) Close() { i.closed = true }
+// Close will close client
+func (i *Client) Close() {
+	i.mutex.Lock()
```
Comment (dim) [inline client.go]: There is no need to lock here, the `i.conn` is never concurrently read/written. Please remove.

File: client_test.go [hunk 2]
```diff
@@ -3,17 +3,42 @@ package redeo
...
+
+func NewMockConn(addr string) *mockConn {
+	return &mockConn{addr: addr}
+}
+
+func (c *mockConn) Close() error {
+	return nil
+}
+
+func (c *mockConn) Network() string {
```
Comment (dim) [inline client_test.go]: Please only expose methods required by the interface.

File: server.go [hunk 3]
```diff
@@ -110,6 +110,15 @@ func (srv *Server) Apply(req *Request) (*Responder, error) {
 	return res, err
 }
 
+func (srv *Server) Close() {
+	srv.mutex.Lock()
```
Comment (dim) [inline server.go]: There is no need to lock here either

**Comment by dim** [comment]
Thanks for the PR, but I think this needs to be implemented slightly differently. One concern is that you are using info for client detection, which is really not the ideal place to track clients. Info is a quick container to store basic information about the server status.

File: client.go (hunk 1 above)
Comment (ngaut) [inline client.go]: There maybe concurrently read/written for i.conn when server close client forcely.

File: client_test.go (hunk 2 above)
Comment (ngaut) [inline client_test.go]: make sence :)

File: client.go (hunk 1 above)
Comment (dim) [inline client.go]: That mutex would not protect it. You would need to use `srv.info.mutex.Lock()` from server.go to lock this. As said in the final statement, tracking clients in info is not a great idea, I will take a look later this week

File: server.go (hunk 3 above)
Comment (ngaut) [inline server.go]: actually, The only way to close server is using Close() method in another thread.

**Comment by ngaut** [comment]
@dim any thoughts?

**Comment by ngaut** [comment]
Thanks

**Pull request closed without merging**